# **XGBoost**

## Importing the libraries

In [104]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the dataset

In [105]:
dataset = pd.read_csv('/kaggle/input/breast-cancer2benign/Data.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

**The Dependant Variable is not in the standard binary form(0,1)**

In [106]:
print(y)

[2 2 2 2 2 4 2 2 2 2 2 2 4 2 4 4 2 2 4 2 4 4 2 2 4 2 2 2 2 2 2 4 2 2 2 4 2
 4 4 4 4 4 4 2 4 2 2 4 4 4 4 4 4 4 4 4 4 4 4 2 4 4 2 4 2 4 4 2 2 4 2 4 4 2
 2 2 2 2 2 2 2 2 4 4 4 4 2 2 2 2 2 2 2 2 2 2 4 4 4 4 2 4 4 4 4 4 2 4 2 4 4
 4 2 2 2 4 2 2 2 2 4 4 4 2 4 2 4 2 2 2 4 2 2 2 2 2 2 2 2 4 2 2 4 2 2 4 2 4
 4 2 2 4 2 2 4 4 2 2 2 2 4 4 2 2 2 2 2 4 4 4 2 4 2 4 2 2 2 4 4 2 4 4 4 2 4
 4 2 2 2 2 2 2 2 2 4 4 2 2 2 4 4 2 2 2 4 4 2 4 4 4 2 2 4 2 2 4 4 4 4 2 4 4
 2 4 4 4 2 4 2 4 4 4 4 2 2 2 2 2 2 4 4 2 2 4 2 4 4 4 2 2 2 2 4 4 4 4 4 2 4
 4 4 2 4 2 4 4 2 2 2 2 4 2 2 4 4 4 4 4 2 4 4 2 2 4 4 2 2 4 4 2 4 2 4 4 2 2
 4 2 2 2 4 2 2 4 4 2 2 4 2 4 2 2 4 2 4 4 4 2 2 4 4 2 4 2 2 4 4 2 2 2 4 2 2
 2 4 4 2 2 2 4 2 2 4 4 4 4 4 4 2 2 2 2 4 4 2 2 2 2 2 2 2 2 2 2 2 2 2 4 2 2
 2 2 4 2 2 2 2 4 2 2 2 2 2 2 2 2 4 2 2 2 2 2 2 2 2 2 2 4 2 4 2 4 2 2 2 2 4
 2 2 2 4 2 4 2 2 2 2 2 2 2 4 4 2 2 2 4 2 2 2 2 2 2 2 2 4 2 2 2 4 2 4 4 4 2
 2 2 2 2 2 2 4 4 4 2 2 2 2 2 2 2 2 2 2 2 4 2 2 4 4 2 2 2 4 4 4 2 4 2 4 2 2
 2 2 2 2 2 2 2 2 2 2 4 2 

##Lable encoding the Dependant Variable

In [107]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y)
print(y)

[0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 1 0 0 1 0 1 1 0 0 1 0 0 0 0 0 0 1 0 0 0 1 0
 1 1 1 1 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 1 0 1 1 0 0 1 0 1 1 0
 0 0 0 0 0 0 0 0 1 1 1 1 0 0 0 0 0 0 0 0 0 0 1 1 1 1 0 1 1 1 1 1 0 1 0 1 1
 1 0 0 0 1 0 0 0 0 1 1 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 1
 1 0 0 1 0 0 1 1 0 0 0 0 1 1 0 0 0 0 0 1 1 1 0 1 0 1 0 0 0 1 1 0 1 1 1 0 1
 1 0 0 0 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 1 1 0 1 1 1 0 0 1 0 0 1 1 1 1 0 1 1
 0 1 1 1 0 1 0 1 1 1 1 0 0 0 0 0 0 1 1 0 0 1 0 1 1 1 0 0 0 0 1 1 1 1 1 0 1
 1 1 0 1 0 1 1 0 0 0 0 1 0 0 1 1 1 1 1 0 1 1 0 0 1 1 0 0 1 1 0 1 0 1 1 0 0
 1 0 0 0 1 0 0 1 1 0 0 1 0 1 0 0 1 0 1 1 1 0 0 1 1 0 1 0 0 1 1 0 0 0 1 0 0
 0 1 1 0 0 0 1 0 0 1 1 1 1 1 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0
 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 0 0 0 0 1
 0 0 0 1 0 1 0 0 0 0 0 0 0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 1 1 0
 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 1 0 0 0 1 1 1 0 1 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 1 0 

## Splitting the dataset into the Training set and Test set

In [108]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

## Feature Extraction using PCA, LDA, Kernel-PCA

**Feature Scaling before Feature Extraction**

In [109]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train_scaled = sc.fit_transform(X_train)
X_test_scaled = sc.transform(X_test)

**Feature Extraction**

Using Principal component analysis
 (PCA)

In [110]:
from sklearn.decomposition import PCA
pca = PCA(n_components = 1)
xg_X_train_PCA = pca.fit_transform(X_train_scaled)
xg_X_test_PCA = pca.transform(X_test_scaled)

Using Linear discriminant analysis (LDA)

In [111]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda = LDA(n_components = 1)
xg_X_train_LDA = lda.fit_transform(X_train_scaled,y_train)
xg_X_test_LDA = lda.transform(X_test_scaled)

Using Kernel-Principal component analysis
 (K-PCA)

In [112]:
from sklearn.decomposition import KernelPCA
kpca = KernelPCA(n_components = 1, kernel = 'rbf')
xg_X_train_kpca = kpca.fit_transform(X_train_scaled)
xg_X_test_kpca = kpca.transform(X_test_scaled)

## Training XGBoost on the Training set

In [113]:
from xgboost import XGBClassifier
classifier1 = XGBClassifier()
classifier2 = XGBClassifier()
classifier3 = XGBClassifier()
classifier4 = XGBClassifier()
classifier5 = XGBClassifier()
classifier1.fit(X_train, y_train)
classifier2.fit(X_train_scaled, y_train)
classifier3.fit(xg_X_train_PCA, y_train)
classifier4.fit(xg_X_train_LDA, y_train)
classifier5.fit(xg_X_train_kpca, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...)

## Making the Confusion Matrix for different cases

In [114]:
from sklearn.metrics import confusion_matrix, accuracy_score
print('For normal Features : ')
y_pred = classifier1.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print('For scaled Features : ')
y_pred = classifier2.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print('For Features Extracted via PCA : ')
y_pred = classifier3.predict(xg_X_test_PCA)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print('For Features Extracted via LDA : ')
y_pred = classifier4.predict(xg_X_test_LDA)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print('For Features Extracted via K-PCA : ')
y_pred = classifier5.predict(xg_X_test_kpca)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))

For normal Features : 
[[85  2]
 [ 1 49]]
0.9781021897810219
For scaled Features : 
[[85  2]
 [ 1 49]]
0.9781021897810219
For Features Extracted via PCA : 
[[83  4]
 [ 1 49]]
0.9635036496350365
For Features Extracted via LDA : 
[[84  3]
 [ 1 49]]
0.9708029197080292
For Features Extracted via K-PCA : 
[[83  4]
 [ 2 48]]
0.9562043795620438


## Applying k-Fold Cross Validation for different cases

In [115]:
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier1, X = X_train, y = y_train, cv = 10)
print('For normal Features : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))
accuracies = cross_val_score(estimator = classifier2, X = X_train_scaled, y = y_train, cv = 10)
print('For scaled Features : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))
accuracies = cross_val_score(estimator = classifier3, X = xg_X_train_PCA, y = y_train, cv = 10)
print('For Features Extracted via PCA : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))
accuracies = cross_val_score(estimator = classifier4, X = xg_X_train_LDA, y = y_train, cv = 10)
print('For Features Extracted via LDA : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))
accuracies = cross_val_score(estimator = classifier5, X = xg_X_train_kpca, y = y_train, cv = 10)
print('For Features Extracted via K-PCA : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))

For normal Features : 
Accuracy: 96.71 %
Standard Deviation: 2.28 %
For scaled Features : 
Accuracy: 96.71 %
Standard Deviation: 2.28 %
For Features Extracted via PCA : 
Accuracy: 96.88 %
Standard Deviation: 2.19 %
For Features Extracted via LDA : 
Accuracy: 97.07 %
Standard Deviation: 1.68 %
For Features Extracted via K-PCA : 
Accuracy: 94.87 %
Standard Deviation: 1.59 %


## **CATBoost**

## Importing the Libraries

In [116]:
pip install catboost

Note: you may need to restart the kernel to use updated packages.


## Importing the Dataset

In [117]:
# Done Above

## Splitting the dataset into training set and test set

In [118]:
# Done Above

## Feature Extraction using PCA, LDA, Kernel-PCA

**Feature Scaling before Feature Extraction**

In [119]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train_scaled = sc.fit_transform(X_train)
X_test_scaled = sc.transform(X_test)

**Feature Extraction**

Using Principal component analysis
 (PCA)

In [120]:
from sklearn.decomposition import PCA
pca2 = PCA(n_components = 1)
cat_X_train_PCA = pca2.fit_transform(X_train_scaled)
cat_X_test_PCA = pca2.transform(X_test_scaled)

Using Linear discriminant analysis (LDA)

In [121]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda2 = LDA(n_components = 1)
cat_X_train_LDA = lda2.fit_transform(X_train_scaled, y_train)
cat_X_test_LDA = lda2.transform(X_test_scaled)

Using Kernel-Principal component analysis
 (K-PCA)

In [122]:
from sklearn.decomposition import KernelPCA
kpca2 = KernelPCA(n_components = 1, kernel = 'rbf')
cat_X_train_kpca = kpca2.fit_transform(X_train_scaled)
cat_X_test_kpca = kpca2.transform(X_test_scaled)

## Training CATBoost on Training Set

In [123]:
from catboost import CatBoostClassifier
classifier1 = CatBoostClassifier(verbose=False)
classifier2 = CatBoostClassifier(verbose=False)
classifier3 = CatBoostClassifier(verbose=False)
classifier4 = CatBoostClassifier(verbose=False)
classifier5 = CatBoostClassifier(verbose=False)
classifier1.fit(X_train, y_train)
classifier2.fit(X_train_scaled, y_train)
classifier3.fit(cat_X_train_PCA, y_train)
classifier4.fit(cat_X_train_LDA, y_train)
classifier5.fit(cat_X_train_kpca, y_train)

## Making the Confusion Matrix

In [124]:
from sklearn.metrics import confusion_matrix, accuracy_score
print('For normal Features : ')
y_pred = classifier1.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print('For scaled Features : ')
y_pred = classifier2.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print('For Features Extracted via PCA : ')
y_pred = classifier3.predict(cat_X_test_PCA)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print('For Features Extracted via LDA : ')
y_pred = classifier4.predict(cat_X_test_LDA)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print('For Features Extracted via K-PCA : ')
y_pred = classifier5.predict(cat_X_test_kpca)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))

For normal Features : 
[[84  3]
 [ 0 50]]
0.9781021897810219
For scaled Features : 
[[84  3]
 [ 0 50]]
0.9781021897810219
For Features Extracted via PCA : 
[[83  4]
 [ 1 49]]
0.9635036496350365
For Features Extracted via LDA : 
[[83  4]
 [ 2 48]]
0.9562043795620438
For Features Extracted via K-PCA : 
[[83  4]
 [ 2 48]]
0.9562043795620438


## Applying K-fold Cross Validation

In [125]:
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier1, X = X_train, y = y_train, cv = 10)
print('For normal Features : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))
accuracies = cross_val_score(estimator = classifier2, X = X_train_scaled, y = y_train, cv = 10)
print('For scaled Features : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))
accuracies = cross_val_score(estimator = classifier3, X = cat_X_train_PCA, y = y_train, cv = 10)
print('For Features Extracted via PCA : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))
accuracies = cross_val_score(estimator = classifier4, X = cat_X_train_LDA, y = y_train, cv = 10)
print('For Features Extracted via LDA : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))
accuracies = cross_val_score(estimator = classifier5, X = cat_X_train_kpca, y = y_train, cv = 10)
print('For Features Extracted via K-PCA : ')
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))

For normal Features : 
Accuracy: 97.26 %
Standard Deviation: 2.03 %
For scaled Features : 
Accuracy: 97.26 %
Standard Deviation: 2.03 %
For Features Extracted via PCA : 
Accuracy: 97.06 %
Standard Deviation: 2.06 %
For Features Extracted via LDA : 
Accuracy: 97.07 %
Standard Deviation: 2.19 %
For Features Extracted via K-PCA : 
Accuracy: 95.78 %
Standard Deviation: 2.19 %


Analyzing the results from XGBoost and CatBoost across different feature engineering techniques provides valuable insights:

1. **Normal Features:**
   - **XGBoost:** 96.71%, CatBoost: 97.26%
   - CatBoost slightly outperforms XGBoost in accuracy on the original feature set.
   - Both models exhibit similar performance with minimal deviation.

2. **Scaled Features:**
   - **XGBoost:** 96.71%, CatBoost: 97.26%
   - Scaled features maintain the same performance as normal features for both models.
   - Feature scaling does not significantly impact model accuracy in this case.

3. **Features Extracted via PCA:**
   - **XGBoost:** 96.88%, CatBoost: 97.06%
   - Both models achieve slightly improved accuracy compared to normal features.
   - PCA demonstrates a positive impact on model performance.

4. **Features Extracted via LDA:**
   - **XGBoost:** 97.07%, CatBoost: 97.07%
   - Both models show enhanced accuracy compared to the original and scaled features.
   - Linear Discriminant Analysis (LDA) is effective in capturing relevant information.

5. **Features Extracted via K-PCA:**
   - **XGBoost:** 94.87%, CatBoost: 95.78%
   - Both models experience a decrease in accuracy compared to other feature sets.
   - Kernel PCA introduces non-linearity that may not benefit the models in this context.

**Conclusion:**
- CatBoost consistently demonstrates slightly higher accuracy than XGBoost across all feature sets.
- Feature scaling has a minimal impact on both models.
- PCA and LDA contribute positively to model accuracy.
- Kernel PCA may not be as effective in this context, resulting in a decrease in accuracy.

The choice between XGBoost and CatBoost depends on the specific requirements of the problem and the significance of computational efficiency, as CatBoost generally performs well "out of the box" without extensive tuning. Consideration of these results aids in selecting the most suitable model and feature engineering strategy for the given dataset.